### ISSUE TITLE
fix: deallocate used resources on services being removed or when the container is reseted to pevent memory leaks

### ISSUE THREAD

**Issue by justinmchase** [typestack/typedi#107]
The remove is too simple, its just dropping references:
https://github.com/typestack/typedi/blob/master/src/ContainerInstance.ts#L218

But to properly manange the life-cycle of an allocated object there needs to be a complementary `dispose()` or close or stop or whatever function on created objects.

The problem arises for me because one of my services uses `setTimeout` which ends up creating a global reference to the instance. There is a `.stop()` function on my service which when called will clear the timeout and allow the object to be free'd but there is nobody calling it.

In production it doesn't matter because I am only allocating all the instances in my container once but in tests I am regularly resetting the Container in the same process between tests. We finally crossed over the bubble where we are running out of memory because of too many un-freed instances.

Instead remove and reset should iterate over all services and call an optional `.dispose()` method if it exists, before dropping references to the service.

Psuedo-code:
```ts
async void reset() {
  await Promise.all(this.services.map(s => s && s.dispose && s.dispose())
  this.services = []
  return this
}
```

Related to #80

**Comment by Spolja** [comment]
Any news about this issue?

**Comment by justinmchase** [comment]
I think someone needs to make a PR, I had a work around and it wasn't affecting production so I haven't had time to do it but the issue remains afaik.

**Comment by NoNameProvided** [comment]
We should check the shape of the service being removed and if we find a `dispose()` function we should call it.

**Comment by NoNameProvided** [comment]
Added in 1938670761ae1f92598c9f6b59f0c69bb144a6d6.

**Comment by NoNameProvided** [comment]
Released as `0.10.0`.
